soroban-env-host/src/budget/dimension.rs

```diff
@@ -72,36 +72,48 @@ impl Debug for BudgetDimension {
 
 impl BudgetDimension {
     pub(crate) fn try_from_config(cost_params: ContractCostParams) -> Result<Self, HostError> {
-        let cost_models = cost_params
-            .0
-            .iter()
-            .map(|p| MeteredCostComponent::try_from(p))
-            .collect::<Result<Vec<MeteredCostComponent>, HostError>>()?;
-
-        Ok(Self {
-            cost_models: cost_models.try_into().map_err(|_| {
-                // error means the cost_params has the wrong dimension
+        let mut bd = BudgetDimension::default();
+        for (i, cp) in cost_params.0.iter().enumerate() {
```

Comment:
I think this implementation is a bit footgun-y. You don't seem to benefit from the fact that you have a vector now and still keep the empty dimensions. Instead of creating default, you should rather be constructing cost models from `cost_params` (i.e. do something like `cost_params.iter().map().collect()`)

Reply:
You are talking about not the implementation but why we keep the full size, even if the input config contains less entries, the non-existent ones are initialized as default?
Keeping the budget internal size constant (as the static size of the xdr) is part of the design and prevents footgun such as forgetting to update an entry (after calibration). There are many places in the code base that checks every cost type existing in the xdr exists logically (e.g. `BudgetImpl::default()`). 
Older protocols with less cost entries will simply touch less model entries. (if that's not the case catch up will fail)

Or if you mean the implementation why it's updating in a loop instead of using collect? I don't think it makes such a big difference. I chose to do it inside a loop to avoid a second allocation call, i.e. vector.resize(default, remaining_count).

Reply:
Sorry for being unclear, I mean that the implementation allocates cost types up to XDR max, then populates the fields. What I was suggesting is to allocate exactly the same number of cost types as was provided in the input. There is also no point in using vector with the approach in this PR. I see your point though - I guess it's better to use a 0 cost than to crash (though if we gracefully handle OOB access, then we should probably be fine as well). FWIW my concern was that we might accidentally access future cost type without noticing.

Reply:
Hmm I see your point, but it just doesn't work very well with the current host design. 
The host is tied to a particular protocol statically -- the current protocol defined by the XDR. The budget structure contains all the cost types that are defined by the XDR, the protocol is runtime gated so that newer cost types are never hit for an older protocol. I don't see much value of adjusting the structure dynamically based on the input config length. It may catch some implementation bug when we write tests against older protocol, but overall it doesn't provide much benefits (those kind of bugs will be caught anyway via replay or other means). 
FIFW making the budget size dynamic requires a bit more restructuring in order to make it consistent between the "construct budget from config" vs "config budget for local testing" workflow, the latter relies on budget::default impl, and changing it to protocol-length dependent it not trivial and makes it harder to work with.
You are right about no need for using vector (dynamic structure) though.

Reply:
I did a few more experiments and decided not to go with the Vector structure (with protocol-dependent size). Don't feel very comfortable changing the invariant that the budget contains all entries defined by the latest protocol.

Addressed your comment about not needing a vector by changing the structure to back to static sized array.